### DISCUSSION TITLE
C# support

### DISCUSSION THREAD

**Discussion by matbeedotcom** [axodox/axodox-machinelearning#6] [General]
Would be curious how to consume this in a C# project- I've added the nuget package but can't quite seem to load the dll's or run a function using P/Invoke

**Comment by axodox** [comment]
You cannot use P/Invoke as that requires C exports function to work (C has a stable binary ABI, but C++ does not).

However there are a number of options for you:
- If you are familiar with C and C++, you could fork this lib and C export functions
  - pros: you can use the same lib, with all features, high performance, 
  - cons: passing arguments such as images through C export functions and PInvoke requires complex marshalling, and it will be cumbersome to use
- You could also use C++/WinRT components to add the functionality to C#
  - pros: all of the above + much simpler to work with than PInvoke, especially with complex objects
  - cons: this will only work on Windows
- Use [this](https://onnxruntime.ai/docs/tutorials/csharp/stable-diffusion-csharp.html) C# example by Microsoft
  - pros: you do not need to work with C++
  - cons: the MS lib has less functionality, e.g. as far as I know, it lacks img2img, inpainting

You could also port this lib to C#, this would however take some time.
